af_ktls.c

```diff
@@ -1486,20 +1547,49 @@ static int tls_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
 	sg_chain(tsk->sgl_send[cnt-1].sg, tsk->sgl_send[cnt-1].npages + 1,
 			tsk->sgtag_send);
 
-	ret = tls_do_encryption(tsk, tsk->sgaad_send, tsk->sg_tx_data, size);
-	if (ret < 0)
+	ret = tls_pre_encrypt(tsk, size);
+	if (ret < 0) {
 		goto send_end;
+	}
 
-	tls_make_prepend(tsk, tsk->header_send, size);
+	order_npages = ret;
+	ret = tls_do_encryption(tsk, tsk->sgaad_send, tsk->sg_tx_data, size);
+	if (ret < 0)
+		goto encrypt_failed;
 
-	ret = kernel_sendmsg(tsk->socket, msg, tsk->vec_send, KTLS_VEC_SIZE,
-			KTLS_RECORD_SIZE(tsk, size));
+	tls_make_prepend(tsk, page_address(tsk->pages_send), size);
+	/*
+	 * The whole record is sent in one go, instead of 3 separate calls
```

Comment:
There is no way to guarantee this with the kernel_sendpage API.  ret returns how many bytes were actually written, and the pages need to hang around until all the data can be sent.

The usual reason for ret returning less than size is because the socket buffer is full, not because there was an error on the socket.